src/algorithms/calorimetry/HEXPLIT.cc

```diff
@@ -0,0 +1,130 @@
+// Copyright (C) 2023 Sebouh Paul
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+// References:
+//   https://arxiv.org/abs/2308.06939
+
+
+#include <DD4hep/Readout.h>
+#include <Evaluator/DD4hepUnits.h>
+#include <TInterpreter.h>
+#include <TInterpreterValue.h>
+#include <edm4hep/Vector3d.h>
+#include <Eigen/Dense>
+#include <fmt/format.h>
+#include <map>
+#include <set>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+#include <iostream>
+
+#include "HEXPLIT.h"
+
+#include "algorithms/calorimetry/HEXPLITConfig.h"
+
+//using namespace edm4eic;
+
+namespace eicrecon {
+
+void HEXPLIT::init(const dd4hep::Detector* detector, std::shared_ptr<spdlog::logger>& logger) {
+    m_log = logger;
+    m_detector = detector;
+}
+
+std::unique_ptr<edm4eic::CalorimeterHitCollection> HEXPLIT::process(const edm4eic::CalorimeterHitCollection &hits){
+
+  double sl=m_cfg.side_length/dd4hep::mm;
+  double layer_spacing=m_cfg.layer_spacing/dd4hep::mm;
+  double MIP=m_cfg.MIP/dd4hep::GeV;
+  double Emin=m_cfg.Emin_in_MIPs*MIP;
+  double tmax=m_cfg.tmax/dd4hep::ns;
+
+  auto volman = m_detector->volumeManager();
+  auto subcellHits = std::make_unique<edm4eic::CalorimeterHitCollection>();
+  double Esum=0;
+  for(const auto& hit : hits){
+    //skip hits that do not pass E and t cuts
+    if (hit.getEnergy()<Emin || hit.getTime()>tmax)
+      continue;
+
+    //keep track of the energy in each neighboring cell
+    std::vector<double> Eneighbors(SUBCELLS, 0.0);
+
+    for (const auto& other_hit : hits){
+      double tol=0.01; //tolerance for rounding errors
+
+      //only look at hits nearby within two layers of the current layer
+      double dz=abs(hit.getLocal().z-other_hit.getLocal().z);
+      if (dz>2.5*layer_spacing || dz<tol)
+        continue;
```

Comment:
If the only reason you use layer_spacing is to check that these are neighbors, then you should probably use the CellID fields. Again, see @veprbl's comments about how this is actually not so different from what other clustering algorithms are doing.

Reply:
In the most recent commit, I changed this so that it uses the getLayer() function to determine which hits are on nearby layers (ie, if the difference in layer number is -2, -1, 1, or 2).  Doing the same for the transverse position is a little more tricky due to the way the row and column numbers works for staggered geometries, but in principle this is possible.

Reply:
> Doing the same for the transverse position is a little more tricky

The LFHCAL had a similar issue due to how the stacking is done, I recall.